File: nnets/mlp/julia/mlpnnets.jl
#
# mlpnnets.jl
# Feed-forward MLP neural network implementation
# Training with backpropagation
#
# Andrei de A. Formiga, 2012-05-07
#

# weights are organized in a 2-dimensional matrix.
# each line in the matrix corresponds to one neuron 
# in the current layer, and contains all the weights 
# arriving in the neuron from the previous layer
type MLPLayer
    n_neurons::Int
    weights::Matrix{Float64}
    outputs::Vector{Float64}

    function MLPLayer(n_neurons::Int, n_neurons_prev::Int)
        new(n_neurons, zeros(Float64, n_neurons, n_neurons_prev+1), zeros(Float64, n_neurons))
    end
end

type MLPNNet
    n_inputs::Int
    layers::Vector{MLPLayer}

    # params: number of inputs, number of neurons per layer 
    function MLPNNet(n_inputs::Int, layers::Vector{Int})
        if length(layers) == 0
            error("MLPNNet: can't create network with 0 layers")
        end
        
        l = Array(MLPLayer, length(layers))
        l[1] = MLPLayer(layers[1], n_inputs)
        for i = 2:length(l)
            l[i] = MLPLayer(layers[i], layers[i-1])
        end
        new(n_inputs, l)
    end
end

get_outputs(nnet::MLPNNet) = nnet.layers[end].outputs

# data and outs have one case per line of each matrix
type DataSet
    n_cases::Int
    n_fields::Int
    n_outputs::Int
    data::Matrix{Float64}
    outs::Matrix{Float64}

    DataSet(n_cases, n_fields, n_outputs) =
        new(n_cases, n_fields, n_outputs, zeros(Float64, n_cases, n_fields),
            zeros(Float64, n_cases, n_outputs))
end

### activation functions
threshold(x::Float64) = x > 0.0? 1.0 : 0.0
sigmoid(x::Float64) = 1.0 / (1.0 + exp(-x))

# y is assumed to be the output of a node with sigmoid activation
dsigmoid(y::Float64) = y * (1.0 - y)  


# initialize weights of the network randomly, with a maximum absolute value
function random_weights(nnet::MLPNNet, maxabs::Float64)
    for l in nnet.layers
        # l.weights = map(x -> (rand() - 0.5) * maxabs * 2, l.weights)
        l.weights = (rand(size(l.weights)) - 0.5) * maxabs * 2
    end
end

# present input to the neural network, calculating outputs for all layers
function present_input(nnet::MLPNNet, input::Vector{Float64}, actf)
end

# vectorized version of present_input
function present_input_vec(nnet::MLPNNet, inputs::Vector{Float64}, actf)
    @assert length(inputs) == nnet.n_inputs
    nnet.layers[1].outputs = map(actf, nnet.layers[1].weights * [1.0, inputs])
    
    for i in 2:length(nnet.layers)
        nnet.layers[i].outputs = map(actf,
                                     nnet.layers[i].weights * [1.0, nnet.layers[i-1].outputs])
    end
    get_outputs(nnet)
end

function batch_train_bprop_vec(nnet::MLPNNet, train_set::DataSet, epochs::Int, 
                               lrate::Float64, actf, dactf)
    # create arrays for deltas (one array per layer, one delta per node)
    deltas = Array(Vector{Float64}, length(nnet.layers))
    for i = 1:length(nnet.layers)
        deltas[i] = Array(Float64, nnet.layers[i].n_neurons)
    end
       
    for ep = 1:epochs
        # create matrices for derivatives (1 matrix per layer, 1 deriv. per weight)
        derivs = Array(Matrix{Float64}, length(nnet.layers))
        for i = 1:length(derivs)
            derivs[i] = zeros(size(nnet.layers[i].weights))
        end

        sse = 0.0
        for i = 1:train_set.n_cases
            # forward propagation
            outs = present_input_vec(nnet, vec(train_set.data[i,:]), actf)
            err = vec(train_set.outs[i,:]) - outs
            sse += sum(err .^ 2)
            
            # deltas for output layer
            deltas[end] = - err .* map(dactf, outs)

            # propagate deltas backwards
            for l = length(nnet.layers)-1:-1:1
                deltas[l] = (nnet.layers[l+1].weights[:, 2:]' * deltas[l+1]) .* map(dactf, nnet.layers[l].outputs)
            end

            # derivatives for first layer
            derivs[1] += deltas[1] * [1.0 train_set.data[i,:]]
            
            # derivatives for remaining layers
            for l = 2:length(nnet.layers)
                derivs[l] += deltas[l] * [1.0 nnet.layers[l-1].outputs']
            end
        end

        if mod(ep, 10) == 0
            println("SSE for epoch: $sse")
        end
      
        # update weights based on derivatives
        for l = 1:length(nnet.layers)
            nnet.layers[l].weights -= derivs[l] .* lrate
        end
    end
end

File: nnets/mlp/mlpnnets.c
/*
 * mlpnnets.c
 * Feed-forward Multi-Layer Perceptron Neural Networks.
 *
 * Andrei de A. Formiga, 2012-03-31
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "mlpnnets.h"

// maximum absolute value of initial weight
#define MAX_ABS_WEIGHT   1.6

// module variable that controls showing of debug information during
// training TODO: implement functions to change it?
int debug_train = 0;

// sigmoid activation function
double sigmoid(double t)
{
    return 1.0 / (1.0 + exp(-t));
}

// derivative of the logistic sigmoid function
// --
// this can be directly calculated
// from the node outputs, which is faster
double dsigmoid(double t)
{
    double sigt = sigmoid(t);
    return sigt * (1.0 - sigt);
}

double threshold(double t)
{
    return (t > 0? 1.0 : 0.0);
}


Layer *create_layer(Layer *prev, int n_neurons)
{
    Layer *res = (Layer*) malloc(sizeof(Layer));

    prev->next = res;

    res->n_neurons = n_neurons;
    res->prev = prev;
    res->a = (double*) malloc(sizeof(double) * n_neurons);
    res->y = (double*) malloc(sizeof(double) * n_neurons);

    // allocate weights (including one for bias for each neuron)
    res->w = (double*) malloc(sizeof(double) * n_neurons * (prev->n_neurons+1));
    res->next = NULL;

    return res;
}

void destroy_layer(Layer *layer)
{
    if (layer != NULL) {
        if (layer->w != NULL)
            free(layer->w);

        if (layer->a != NULL)
            free(layer->a);

        if (layer->y != NULL)
            free(layer->y);

        free(layer);
    }
}


Network *create_network(int n_inputs)
{
    Network *res = (Network*) malloc(sizeof(Network));

    res->n_layers = 1;
    res->input_layer = (Layer*) malloc(sizeof(Layer));
    res->input_layer->n_neurons = n_inputs;
    res->input_layer->w = NULL;
    res->input_layer->prev = NULL;
    res->input_layer->next = NULL;
    res->input_layer->a = NULL;
    res->input_layer->y = (double*) malloc(sizeof(double) * n_inputs);

    res->output_layer = res->input_layer;

    return res;
}

Layer* add_layer(Network *nnet, int n_neurons)
{
    Layer *new_layer = create_layer(nnet->output_layer, n_neurons);
    nnet->output_layer = new_layer;
    nnet->n_layers++;

    return new_layer;
}

void destroy_network(Network *nnet)
{
    Layer *l1, *l2;

    l1 = nnet->input_layer;

    while (l1 != NULL) {
        l2 = l1->next;
        destroy_layer(l1);
        l1 = l2;
    }

    free(nnet);
}

void print_network_structure(Network *nnet)
{
    int   i, j;
    Layer *l = nnet->input_layer->next;

    printf("--- Neural network structure ---\n");
    printf("Layers: %d\n", nnet->n_layers);
    printf("Inputs: %d\n", nnet->input_layer->n_neurons);
    printf("Outputs: %d\n", nnet->output_layer->n_neurons);

    i = 0;
    while (l != NULL) {
        printf("Weights[%d]: ", i);
        for (j = 0; j < l->n_neurons * (l->prev->n_neurons+1); ++j)
            printf("%2.1f ", l->w[j]);
        printf("\n");
        l = l->next;
        ++i;
    }

    printf("--------------------------------\n");
}

void initialize_weights(Network *nnet, unsigned int seed)
{
    int r;
    int i, j;
    Layer *l = nnet->input_layer->next;

    srand(seed);

    while (l != NULL) {
        for (i = 0; i < l->n_neurons; ++i)
            for (j = 0; j < l->prev->n_neurons+1; ++j) {
                r = rand();
                W(l, i, j) = (r - (RAND_MAX / 2.0)) *
                             (2.0 * MAX_ABS_WEIGHT / RAND_MAX);
            }
        l = l->next;
    }
}

// propagates the input forward, calculating the network outputs
// activf is the activation function to use on the outputs
// (assumes input is the same size as the number of neurons in input layer)
void forward_prop(Network *nnet, double (*activf)(double), double *input)
{
    int   i, j;
    Layer *prev_layer = nnet->input_layer;
    Layer *layer = prev_layer->next;

    // copy inputs to the input layer neurons
    for (i = 0; i < nnet->input_layer->n_neurons; ++i)
        nnet->input_layer->y[i] = input[i];

    while (layer != NULL) {
        for (i = 0; i < layer->n_neurons; ++i) {
            // compute the bias
            layer->a[i] = W(layer, i, 0) * 1.0;
            // add weights * inputs
            for (j = 1; j < (prev_layer->n_neurons+1); ++j)
                layer->a[i] += W(layer, i, j) * prev_layer->y[j-1];
            // apply activation function
            layer->y[i] = activf(layer->a[i]);
        }
        prev_layer = layer;
        layer = prev_layer->next;
    }
}

double **create_weight_derivatives_matrix(Network *nnet)
{
    int    i;
    int    n_weights;  // number of weights per layer
    double **deriv;
    Layer  *l = nnet->input_layer->next;

    deriv = (double**) malloc(sizeof(double*) * (nnet->n_layers-1));
    if (deriv == NULL)
        return NULL;

    for (i = 0; i < nnet->n_layers-1; ++i, l = l->next) {
        n_weights = l->n_neurons * (l->prev->n_neurons+1);
        deriv[i] = (double*) calloc(n_weights, sizeof(double));
        if (deriv[i] == NULL)
            return NULL;
    }

    return deriv;
}

// create a matrix for deltas
// (could reduce space to keep deltas for only 2 layers at a time)
double **create_delta_matrix(Network *nnet)
{
    int    i;
    double **deltas;
    Layer  *l = nnet->input_layer->next;

    deltas = (double**) malloc(sizeof(double*) * (nnet->n_layers-1));
    if (deltas == NULL)
        return NULL;

    for (i = 0; i < nnet->n_layers-1; ++i, l = l->next) {
        deltas[i] = (double*) malloc(sizeof(double) * l->n_neurons);
        if (deltas[i] == NULL)
            return NULL;
    }

    return deltas;
}

// calculate deltas for the nodes given the desired outputs
// return the sum of square errors for these outputs
double calculate_deltas(Network *nnet, double **delta, double *d_output,
                        double (*dactf)(double))
{
    int    i, j;
    int    ln;     // layer number
    double d, err = 0.0;
    Layer  *l = nnet->output_layer;

    // calculate deltas for output layer
    ln = nnet->n_layers-2;
    for (i = 0; i < l->n_neurons; ++i) {
        d = d_output[i] - l->y[i];
        err += d * d;
        delta[ln][i] = -d * dactf(l->a[i]);
    }

    // calculate deltas for hidden layers
    for (ln--, l=l->prev; ln >=0; ln--, l=l->prev) {
        for (i = 0; i < l->n_neurons; ++i) {
            delta[ln][i] = 0.0;
            for (j = 0; j < l->next->n_neurons; ++j)
                // use i+1-th weight to compensate for bias node
                delta[ln][i] += W(l->next, j, i+1) * delta[ln+1][j];
            delta[ln][i] *= dactf(l->a[i]);
        }
    }

    return err;
}

// given deltas for the nodes,
// calculate the derivatives in relation to the weights
void calculate_derivatives(Network *nnet, double **delta, double **deriv)
{
    int   ln;   // layer number
    int   nn;   // neuron (node) number
    int   wn;   // weight number
    int   ws;   // number of weights per node in layer
    Layer *l = nnet->input_layer->next;

    for (ln = 0; ln < nnet->n_layers-1; ++ln, l=l->next) {
        ws = l->prev->n_neurons + 1;
        for (nn = 0; nn < l->n_neurons; ++nn) {
            // dE_p/dw_{n0}^l = delta_n^l (bias weight)
            deriv[ln][nn*ws] += delta[ln][nn];
            for (wn = 1; wn < ws; ++wn)
                // dE_p/dw_{nw}^l = delta_n^l * y_w^{l-1}
                deriv[ln][nn*ws+wn] += delta[ln][nn] * l->prev->y[wn-1];
        }
    }
}

// train network nnet in batch mode with backpropagation,
// using the provided dataset, for a number of epochs,
// using lrate as the learning rate and actf as
// activation function (dactf is its derivative)
// ---
// return an approximation to the
// final sum-of-squares error for network after training
double batch_train(Network *nnet, DataSet *dset, double lrate, int epochs,
                   double (*actf)(double), double (*dactf)(double))
{
    int    i, j;
    int    ln;
    int    wn, ws;
    double err;
    double **deriv;   // per-weight derivatives
    double **delta;   // per-node delta for derivative calculation
    Layer  *l;

    // check to see if the dataset matches the network
    if (dset->input_size != nnet->input_layer->n_neurons) {
        fprintf(stderr, "batch_train: Size of input in dataset is different \
                         from input layer in neural net\n");
        return 0.0;
    }

    if (dset->output_size != nnet->output_layer->n_neurons) {
        fprintf(stderr, "batch_train: Size of output in dataset is different \
                         from output layer in neural net\n");
        return 0.0;
    }

    deriv = create_weight_derivatives_matrix(nnet);
    delta = create_delta_matrix(nnet);
    // TODO: verify if allocations failed

    for (j = 0; j < epochs; ++j) {
        err = 0.0;
        // clear the deriv matrix
        l = nnet->input_layer->next;
        for (ln = 0; ln < nnet->n_layers-1; ++ln, l=l->next) {
            for (i = 0; i < l->n_neurons * (l->prev->n_neurons+1); ++i)
                deriv[ln][i] = 0.0;
        }

        for (i = 0; i < dset->n_cases; ++i) {
            // propagate forward and calculate network outputs
            forward_prop(nnet, actf, dset->input[i]);

            // calculate deltas for nodes
            err += calculate_deltas(nnet, delta, dset->output[i], dactf);

            // calculate error derivatives for the current input
            calculate_derivatives(nnet, delta, deriv);
        }

        // now derivatives for this epoch have been calculated
        // update weights based on derivatives
        l = nnet->input_layer->next;
        for (ln = 0; ln < nnet->n_layers-1; ++ln, l=l->next) {
            ws = l->prev->n_neurons + 1;
            for (i = 0; i < l->n_neurons; ++i)
                for (wn = 0; wn < ws; ++wn)
                    W(l, i, wn) -= lrate * deriv[ln][i*ws+wn];
        }

        if (debug_train)
            printf("Epoch: %d - Total SSE for epoch: %6.4f\n", j, err);
    }

    return err;
}

// allocate dataset arrays based on the number of cases
// and input/output sizes
void allocate_dataset_arrays(DataSet *dset)
{
    int i;

    // TODO: verify allocation results
    dset->input = (double**) malloc(sizeof(double*) * dset->n_cases);
    dset->output = (double**) malloc(sizeof(double*) * dset->n_cases);

    for (i = 0; i < dset->n_cases; ++i) {
        dset->input[i] = (double*) malloc(sizeof(double) * dset->input_size);
        dset->output[i] = (double*) malloc(sizeof(double) * dset->output_size);
    }
}

// free the double arrays in an array of arrays
void free_arrayarray(double **array, int rows)
{
    int i;

    for (i = 0; i < rows; ++i)
        if (array[i] != NULL)
            free(array[i]);
}

// free the memory occupied by a dataset
void free_dataset(DataSet *dset)
{
    if (dset->input != NULL) {
        free_arrayarray(dset->input, dset->n_cases);
        free(dset->input);
    }

    if (dset->output != NULL) {
        free_arrayarray(dset->output, dset->n_cases);
        free(dset->output);
    }

    free(dset);
}
